### PR TITLE
ceph: allow replicated/erasure pool specs = null

### PR DESCRIPTION
**Pull request by BlaineEXE**
Allow `replicated` or `erasureCoded` `pool` spec to be null. This allows
us to still set a min on `replicated.size` in the CRD as well as mins on
`erasureCoded.codingChunks` and `erasureCoded.dataChunks`.

Signed-off-by: Blaine Gardner <blaine.gardner@redhat.com>

**Description of your changes:**

**Which issue is resolved by this Pull Request:**
Resolves #

**Checklist:**

- [ ] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://rook.io/docs/rook/master/development-flow.html#commit-structure).
- [ ] **Skip Tests for Docs**: Add the flag for skipping the build if this is only a documentation change. See [here](https://github.com/rook/rook/blob/master/INSTALL.md#skip-ci) for the flag.
- [ ] **Skip Unrelated Tests**: Add a flag to run tests for a specific storage provider. See [test options](https://github.com/rook/rook/blob/master/INSTALL.md#test-storage-provider).
- [ ] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/master/development-flow.html#submitting-a-pull-request)
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.
- [ ] Pending release notes updated with breaking and/or notable changes, if necessary.
- [ ] Upgrade from previous release is tested and upgrade user guide is updated, if necessary.
- [ ] Code generation (`make codegen`) has been run to update object specifications, if necessary.

### CODE REVIEW

File: build/codegen/codegen.sh [hunk 1]
```diff
@@ -22,6 +22,11 @@ codegendir="${scriptdir}/../../vendor/k8s.io/code-generator"
 # ensures the vendor dir has the right deps, e,g. code-generator
 go mod vendor
 
+# Set these so that we don't 'go get' cross-compiled binaries
+GOOS="$(go env GOHOSTOS)"
+GOARCH="$(go env GOHOSTARCH)"
+export CGO_ENABLED=0
```
Comment (BlaineEXE) [inline build/codegen/codegen.sh:28]: Allow `make codegen` on mac.

Comment (travisn) [reply build/codegen/codegen.sh:28]: Separate PR needed? though I'm able to run `make codegen` on mac without this...

**Review by travisn — CHANGES_REQUESTED** [review]
While this is a nice to have and the right approach, I wonder if upgraded clusters will have issues with the schema restrictions. We have allowed 0 in the schema for a looong time. For example, if you start with a pre-1.16 v1beta1 pool CRD, then upgrade to a v1 CRD, will everything work? Or what about the few people who might have 0 in their CRs? Will they be ok for upgrades? It may be that only newly created pools are rejected, but it will need some manual testing.

File: build/codegen/codegen.sh (hunk 1 above)
Comment (leseb) [reply build/codegen/codegen.sh:28]: +1 I don't need that to run codegen.

File: pkg/apis/ceph.rook.io/v1/types.go [hunk 2]
```diff
@@ -587,11 +587,13 @@ type PoolSpec struct {
 
 	// The replication settings
 	// +optional
-	Replicated ReplicatedSpec `json:"replicated,omitempty"`
+	// +nullable
+	Replicated *ReplicatedSpec `json:"replicated,omitempty"`
```
Comment (leseb) [inline pkg/apis/ceph.rook.io/v1/types.go:591]: If it's a pointer I don't think you need the `// +nullable`.

**Comment by BlaineEXE** [comment]
> While this is a nice to have and the right approach, I wonder if upgraded clusters will have issues with the schema restrictions. We have allowed 0 in the schema for a looong time. For example, if you start with a pre-1.16 v1beta1 pool CRD, then upgrade to a v1 CRD, will everything work? Or what about the few people who might have 0 in their CRs? Will they be ok for upgrades? It may be that only newly created pools are rejected, but it will need some manual testing.

I agree. I was also thinking maybe this makes sense to leave for 1.7 so we don't risk breaking things in a 1.6.z.

**Comment by BlaineEXE** [comment]
Found out today that `make codegen` won't work if `rook` isn't in `~/go/src/github.com/rook/`. Seems to have something to do with the `--output-base` flag to the generators and how they expect go code to be laid out.

File: pkg/apis/ceph.rook.io/v1/types.go (hunk 2 above)
Comment (BlaineEXE) [reply pkg/apis/ceph.rook.io/v1/types.go:591]: Are you sure? If I don't have this in, the schema doesn't generate `nullable: true` which I think is needed for allowing Rook to actually get `null` values for those.

**Pull request closed without merging**